crates/orderbook/src/cow_subsidy.rs

```diff
@@ -24,67 +24,114 @@ impl Default for FixedCowSubsidy {
 
 #[async_trait::async_trait]
 impl CowSubsidy for FixedCowSubsidy {
+    /// On success returns a factor of how much of the usual fee an account has to pay based on the
+    /// amount of COW tokens it owns.
     async fn cow_subsidy_factor(&self, _: H160) -> Result<f64> {
         Ok(self.0)
     }
 }
 
-#[derive(Debug, Clone, Copy)]
-enum SubsidyTier {
-    None,
-    Full,
+#[derive(Copy, Clone, Debug)]
+pub struct SubsidyTier {
+    /// How many base units of COW someone needs to own to qualify for this subsidy tier.
+    threshold: U256,
+    /// How much of the usual fee a user in this tier has to pay.
+    fee_factor: f64,
+}
+
+impl std::str::FromStr for SubsidyTier {
+    type Err = anyhow::Error;
+    fn from_str(tier: &str) -> Result<Self, Self::Err> {
+        let mut parts = tier.split(':');
+        let threshold = parts.next().ok_or(anyhow::anyhow!("missing threshold"))?;
+        let threshold: f64 = threshold
+            .parse()
+            .map_err(|_| anyhow::anyhow!("can not parse threshold {} as f64", threshold))?;
+
+        let fee_factor = parts.next().ok_or(anyhow::anyhow!("missing fee factor"))?;
+        let fee_factor: f64 = fee_factor
+            .parse()
+            .map_err(|_| anyhow::anyhow!("can not parse fee factor {} as f64", fee_factor))?;
+
+        if parts.next().is_some() {
+            anyhow::bail!("too many arguments for subsidy tier");
+        }
+
+        Ok(SubsidyTier::new(
+            U256::from_f64_lossy(1e18 * threshold),
+            fee_factor,
+        ))
+    }
+}
+
+impl SubsidyTier {
+    pub fn new(threshold: U256, fee_factor: f64) -> Self {
+        assert!(fee_factor <= 1.0);
+        assert!(fee_factor >= 0.0);
+        Self {
+            threshold,
+            fee_factor,
+        }
+    }
 }
 
 pub struct CowSubsidyImpl {
     cow_token: ERC20,
-    threshold: U256,
-    factor: f64,
-    cache: Mutex<TimedSizedCache<H160, SubsidyTier>>,
+    // sorted by threshold in increasing order, no duplicated thresholds
+    subsidy_tiers: Vec<SubsidyTier>,
```

Comment:
Maybe a `BTreeMap<Threshold, Factor>` might be appropriate here?

Reply:
I thought since we will never lookup a key but instead always loop through the keys a map would be weird.
Don't have a strong opinion, though.

Reply:
> never lookup a key but instead always loop through the keys a map would be weird

It allows you to binary search for the key. [For example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=67bc32d9559e6905dafd1bbe270062ba):
```rust
let tier = map.range(..balance).rev().next();
```

Reply:
`BTreeMap::range` seems to work. In that case the reasons for the map outweigh the one against. 👍